Add async flag and functionality to gaiacli send

davekaj opened the pull request:
<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺ 
v                               ✰  Thanks for creating a PR! ✰    
v    Before smashing the submit button please review the checkboxes. 
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  --> 

* [x] ~~Updated all relevant documentation in docs~~ didnt see anywhere to add it 
* [x] Updated all code comments where relevant
* [x] ~~Wrote tests~~  - no new funcs to test, only cli
* [x] Updated CHANGELOG.md
* [x] Updated Gaia/Examples
* [x] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

This PR addresses https://github.com/cosmos/cosmos-sdk/issues/1193.

It adds an async flag to gaiacli send, to allow for many async txs in the same block. 

For now, sequence will have to be supplied from a script with --sequence if you want to send a lot of txs very quickly. Otherwise it will grab the first sequence from tendermint, but then it will continue to try the same sequence number. This can be updated once there is a way to store sequences locally (see comments in #1193) 

I don't think tests are needed, since this is just a cli call that uses funcs that already exist. I have tested it on a local testnet with branch develop and it works. 

Once you guys approve the changes, I will add in the async flag to democoin and basecoin, update changelog, and see if any docs need to be changed. Then it should be mergable @ValarDragon @cwgoes

davekaj commented:
check out this gist to help with testing it out in the cli 

https://gist.github.com/davekaj/4ec2d1a33004708f5b5db5a002862557

ValarDragon commented on client/context/helpers.go:
```diff
@@ -185,6 +201,32 @@ func (ctx CoreContext) EnsureSignBuildBroadcast(name string, msgs []sdk.Msg, cdc
 
+// sign and build the transaction from the msg
+func (ctx CoreContext) EnsureSignBuildBroadcastAsync(name string, msgs []sdk.Msg, cdc *wire.Codec) (res *ctypes.ResultBroadcastTx, err error) {
+
```
It looks like this is the same as EnsureSignBuildBroadcast, could we abstract the functionality of these two into another method, perhaps `ensureSignBuild` which takes the same parameters as these two, but returns `txBytes, err`, and if that doesn't error, then run the respective broadcast method on each? (Just to reduce code duplication. I know this will be changed whenever we change how we're getting sequence numbers for async tx's, but since async is primarily intended for testing, I think that will be awhile)

ValarDragon commented on client/context/viper.go:
```diff
@@ -56,7 +56,7 @@ func defaultChainID() (string, error) {
 }
 
-// EnsureSequence - automatically set sequence number if none provided
+// EnsureAccount - automatically set account number if none provided
```
Thanks for catching this!

davekaj replied:
np

ValarDragon commented on x/bank/client/cli/sendtx.go:
```diff
@@ -47,16 +48,28 @@ func SendTxCmd(cdc *wire.Codec) *cobra.Command {
+				if err != nil {
+					return err
+				}
+				fmt.Println("Async tx send. tx hash: ", res.Hash)
```
Should this be `Async tx sent`? (`sent` instead `send)

rigelrozanski replied:
meh

davekaj commented:
I updated the EnsureSignBuild tx as requested. it required me to go and update about 8 other instances where EnsureSignBuildBroadcast was being called. However it all works now, and I updated changelog, and i realized that democoin and basecoin already inherit the sendTxCmd, so they have the --async flag too. I think it should be ready to merge

ValarDragon commented on client/context/helpers.go:
```diff
@@ -160,7 +176,7 @@ func (ctx CoreContext) SignAndBuild(name, passphrase string, msgs []sdk.Msg, cdc
 
 // sign and build the transaction from the msg
-func (ctx CoreContext) EnsureSignBuildBroadcast(name string, msgs []sdk.Msg, cdc *wire.Codec) (res *ctypes.ResultBroadcastTxCommit, err error) {
+func (ctx CoreContext) EnsureSignBuild(name string, msgs []sdk.Msg, cdc *wire.Codec) (txBytes []byte, err error) {
```
Sorry, this is my fault for being insufficiently clear in my last message. I meant could we have ensureSignBuild as a private method which both EnsureSignBuildBroadcast and the async version call. Removing the broadcasting here makes it more work (as you've been changing) in all of the clients. 

Could you bring back EnsureSignBuildBroadcast? Sorry about the confusion / extra work

davekaj replied:
yep i can do that it will be easy

davekaj commented:
yeah i think i remember the most I was getting at any point was 3-4 txs per block.  but the blocks were coming to consensus so quickly i think that was why
